staging/src/k8s.io/sample-controller/controller_test.go

```diff
@@ -0,0 +1,271 @@
+			break
+		}
+
+		expectedAction := f.kubeactions[i]
+		if !(expectedAction.Matches(action.GetVerb(), action.GetResource().Resource) && action.GetSubresource() == expectedAction.GetSubresource()) {
+			f.t.Errorf("Expected\n\t%#v\ngot\n\t%#v", expectedAction, action)
+			continue
+		}
+	}
+
+	if len(f.kubeactions) > len(k8sActions) {
+		f.t.Errorf("%d additional expected actions:%+v", len(f.kubeactions)-len(k8sActions), f.kubeactions[len(k8sActions):])
+	}
+}
+
+func filterInformerActions(actions []core.Action) []core.Action {
+	ret := []core.Action{}
+	for _, action := range actions {
+		if len(action.GetNamespace()) == 0 &&
+			(action.Matches("list", "foos") ||
+				action.Matches("watch", "foos") ||
+				action.Matches("list", "deployments") ||
+				action.Matches("watch", "deployments")) {
+			continue
+		}
+		ret = append(ret, action)
+	}
+
+	return ret
+}
+
+func (f *fixture) expectCreateDeploymentAction(d *apps.Deployment) {
+	f.kubeactions = append(f.kubeactions, core.NewCreateAction(schema.GroupVersionResource{Resource: "deployments"}, d.Namespace, d))
+}
+
+func (f *fixture) expectUpdateDeploymentAction(d *apps.Deployment) {
+	f.kubeactions = append(f.kubeactions, core.NewUpdateAction(schema.GroupVersionResource{Resource: "deployments"}, d.Namespace, d))
+}
+
+func (f *fixture) expectUpdateFooStatusAction(foo *samplecontroller.Foo) {
+	action := core.NewUpdateAction(schema.GroupVersionResource{Resource: "foos"}, foo.Namespace, foo)
+	// TODO: Until #38113 is merged, we can't use Subresource
+	//action.Subresource = "status"
+	f.actions = append(f.actions, action)
+}
+
+func getKey(foo *samplecontroller.Foo, t *testing.T) string {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(foo)
+	if err != nil {
+		t.Errorf("Unexpected error getting key for foo %v: %v", foo.Name, err)
+		return ""
+	}
+	return key
+}
+
+func TestCreatesDeployment(t *testing.T) {
+	f := newFixture(t)
+	foo := newFoo("test", int32Ptr(1))
+
+	f.fooLister = append(f.fooLister, foo)
+	f.objects = append(f.objects, foo)
+
+	d := newDeployment(foo)
+	f.expectCreateDeploymentAction(d)
+	f.expectUpdateFooStatusAction(foo)
+
+	f.run(getKey(foo, t))
+}
+
+func TestDoNothing(t *testing.T) {
+	f := newFixture(t)
+	foo := newFoo("test", int32Ptr(1))
+	d := newDeployment(foo)
+
+	f.fooLister = append(f.fooLister, foo)
+	f.objects = append(f.objects, foo)
+	f.deploymentLister = append(f.deploymentLister, d)
+	f.kubeobjects = append(f.kubeobjects, d)
+
+	f.expectUpdateFooStatusAction(foo)
+	f.run(getKey(foo, t))
+}
+
+func TestUpdateDeployment(t *testing.T) {
+	f := newFixture(t)
+	foo := newFoo("test", int32Ptr(1))
+	d := newDeployment(foo)
+
+	// Update replicas
+	foo.Spec.Replicas = int32Ptr(2)
+
+	f.fooLister = append(f.fooLister, foo)
+	f.objects = append(f.objects, foo)
+	f.deploymentLister = append(f.deploymentLister, d)
+	f.kubeobjects = append(f.kubeobjects, d)
+
+	f.expectUpdateFooStatusAction(foo)
+	f.expectUpdateDeploymentAction(d)
+	f.run(getKey(foo, t))
+}
```

Comment:
From what I can see, this test doesn't actually verify that the deployment has been updated to have 2 replicas, only that the deployment was updated in *some way*. Is that desired?

Reply:
This enough for internal k8s controllers, but let me check if we can do better job here.

Reply:
Ok, I figured out the better way, we can cast action objects and call .GetObject() to receive actual object and compare it with the expected one.

Only one problem, I'm on  vacation next week and will not have time to work on it.

So we have two options here:
1. Merge it as it is, and I will submit better implementation in one week
2. Wait one week for it.

Reply:
Done